README.md
**EctoStreamFactory** helps to utilize [StreamData generators](https://hexdocs.pm/stream_data/ExUnitProperties.html#gen/1)
as [Ecto factories](https://hexdocs.pm/ecto/test-factories.html).

You can define one factory and use it in the following scenarios:
* Regular unit/integration/acceptance tests
* Property-based tests
* Load/stress/performance tests
* Database seeding

[HexDocs](https://hexdocs.pm/ecto_stream_factory)

You can read about property-based testing in Fred Hebert's [book](https://pragprog.com/titles/fhproper/property-based-testing-with-proper-erlang-and-elixir).

## Installation

Add EctoStreamFactory dependency to `mix.exs`:

```elixir
def deps do
  [
    {:ecto_stream_factory, "~> 0.2", only: [:test, :dev]}
  ]
end
```

Add `:stream_data` to your `.formatter.exs`:

```elixir
[
  import_deps: [:stream_data]
]
```

Create a factory module in `test/support/factory.ex`. Generator functions should have "_generator" suffix:

```elixir
defmodule MyApp.Factory do
  use EctoStreamFactory, repo: MyApp.Repo

  def user_generator do
    gen all name <- string(:alphanumeric, min_length: 1),
            age <- integer(15..80),
            email <- email_generator() do
      %MyApp.User{name: name, age: age, email: email}
    end
  end

  def post_generator do
    gen all author <- user_generator(),
            body <- string(:alphanumeric, min_length: 10) do
      %MyApp.Post{author: author, body: body}
    end
  end

  def email_generator do
    gen all username <- string(:alphanumeric, min_length: 1),
            domain <- member_of(~w(gmail.com protonmail.com yandex.com)) do
      "#{username}#{System.unique_integer([:positive, :monotonic])}@#{domain}"
    end
  end

  # handy for controller tests
  def user_params_generator do
    gen all user <- user_generator(),
            # you can also use Faker to generate human-friendly data
            last_name <- constant(Faker.Person.last_name()),
            idempotency_key <- constant(Ecto.UUID.generate()) do
      %{
        "name" => user.name,
        "last_name" => last_name,
        "age" => to_string(user.age),
        "email" => user.email,
        "idempotency_key" => idempotency_key
      }
    end
  end
end
```

Make sure that `mix.exs` is configured to compile the factory for required environments:

```elixir
def project do
  [
    elixirc_paths: elixirc_paths(Mix.env())
  ]
end

defp elixirc_paths(env) when env in [:test, :dev], do: ["lib", "test/support"]
defp elixirc_paths(_), do: ["lib"]
```

Optionally import the factory in `.iex.exs` to simplify its usage inside `iex -S mix` console on your development machine:

```elixir
import MyApp.Factory
```

For a vanilla Elixir project you can import the factory in every test module with [ExUnti.CaseTemplate](https://hexdocs.pm/ex_unit/ExUnit.CaseTemplate.html).
Create a file `test/support/case.ex` and add:
```elixir
defmodule MyApp.Case do
  use ExUnit.CaseTemplate

  using do
    quote do
      import MyApp.Factory
    end
  end
end
```
Then in your test files replase `use ExUnit.Case` with `use MyApp.Case`

For a Phoenix project you can import the factory in `support/data_case.ex`, `support/conn_case.ex`, `support/channel_case`:

```elixir
defmodule MyAppWeb.ConnCase do
  using do
    quote do
      ...
      import MyApp.Factory
    end
  end
end
```

## Usage in regular tests

```elixir
iex> build(:email)
"S1@protonmail.com"

iex> build(:user)
%User{id: nil, name: "a", age: 33, email: "I2@yandex.com"}

iex> build(:user, name: "Bob")
%User{id: nil, name: "Bob", age: 28, email: "S3@gmail.com"}

iex> build(:post, text: "Hello world")
%Post{
  id: nil,
  text: "Hello world",
  author: %User{id: nil, name: "b", age: 28, email: "l4@gmail.com"}
}

iex> build(:user_params, %{"idempotency_key" => "123", "new_param" => "foo"})
%{
  "name" => "2BO",
  "last_name" => "Herman",
  "age" => "32",
  "email" => "a@gmail.com",
  "idempotency_key" => "123",
  "new_params" => "foo"
}

iex> build_list(2, :user, name: fn n -> "user#{n}" end)
[
  %User{id: nil, name: "user1", age: 51, email: "n5@gmail.com"},
  %User{id: nil, name: "user2", age: 40, email: "O6@yandex.com"}
]

iex> insert!(:user)
%User{id: 1, name: "b", age: 23, email: "az7@gmail.com"}

iex> insert!(:user, gender: "female")
** (EctoStreamFactory.MissingKeyError) MyApp.Factory.user_generator does not generate :gender field.

iex> insert(:user, [email: "az7@gmail.com"], on_conflict: :nothing)
%User{id: nil, name: "c", age: 44, email: "az7@gmail.com"}

iex> insert(:post, author: build(:user, name: "Jane"))
%Post{
  id: 1,
  text: "kjfwi245lfh",
  author: %User{id: 2, name: "Jane", age: 34, email: "jhg8@yandex.com"}
}

iex> insert_list(2, :user, age: 18)
[
  %User{id: 3, name: "bc", age: 18, email: "kl9@protonmail.com"},
  %User{id: 4, name: "bd", age: 18, email: "hj10@yandex.com"}
]
```

## Usage in property-based tests

```elixir
defmodule MyAppTest do
  use MyApp.Case, async: true
  use ExUnitProperties

  describe "user properties" do
    property "contact info contains user name and email" do
      check all user <- user_generator() do
        info = MyApp.User.contact_info(user) 
        assert String.starts_with?(info, user.name)
        assert info =~ user.email
      end
    end

    property "adult users older than 18" do
      check all user <- user_generator() do
        assert MyApp.User.adult?(user) == user.age >= 18
      end
    end
  end
end
```

## Usage in seeding the database

In `priv/repo/seeds.exs`

```elixir
import MyApp.Factory

insert_list(100, :post)
```
Then run it with `mix run priv/repo/seeds.exs`

## Advanced cases
### How to conditionally generate values based on the result of previous generators?

```elixir
def user_generator do
  gen all language_code <- member_of(~w(ru en)),
          last_name <- user_last_name(language_code) do
    %User{
      language_code: language_code,
      last_name: last_name
    }
  end
end

defp user_last_name("ru"), do: member_of(~w(Ivanov Petrov))
defp user_last_name("en"), do: member_of(~w(Smith Brown))
```

### How to modify the output of a generator inside another generator?

```elixir
def admin_generator do
  gen all admin <-
            bind(user_generator(), fn user ->
              Map.put(user, :type, "admin")
            end) do
    admin
  end
end
```


CHANGELOG.md
# Changelog

## v0.2.2 (2022-05-30)

### Enhancements

  * Improve documentation
  * Add more tests

### Bug fix

  * Convert `t:EctoStreamFactory.overwrites/0` to a map to prevent bugs

## v0.2.1 (2022-05-28)

### Enhancements

  * Improve documentation
  * Update dependencies

## v0.2.0 (2021-04-24)

### Enhancements

  * Add `build!/2`
  * Add `build_list!/3`
  * Add `insert!/3`
  * Add `insert_list!/4`

All the new functions raise `EctoStreamFactory.MissingKeyError` error if a generated struct, map or keyword list cannot be fully merged with the provided attributes.


config/test.exs
import Config

config :ecto_stream_factory, ecto_repos: [EctoStreamFactory.Repo]

config :ecto_stream_factory, EctoStreamFactory.Repo,
  hostname: "localhost",
  database: "ecto_stream_factory_test",
  username: "postgres",
  password: "postgres",
  pool: Ecto.Adapters.SQL.Sandbox

config :logger, level: :info


test/support/repo.ex
defmodule EctoStreamFactory.Repo do
  use Ecto.Repo,
    otp_app: :ecto_stream_factory,
    adapter: Ecto.Adapters.Postgres
end


test/test_helper.exs
{:ok, _pid} = EctoStreamFactory.Repo.start_link()
Ecto.Adapters.SQL.Sandbox.mode(EctoStreamFactory.Repo, :manual)

ExUnit.start()


mix.exs
defmodule EctoStreamFactory.MixProject do
  use Mix.Project

  @project_url "https://github.com/ibarchenkov/ecto_stream_factory"
  @version "0.2.2"

  def project do
    [
      app: :ecto_stream_factory,
      version: @version,
      elixir: ">= 1.10.0",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      elixirc_paths: elixirc_paths(Mix.env()),
      aliases: aliases(),

      # Hex
      description: "A factory library for property-based and regular tests",
      package: package(),

      # Docs
      name: "EctoStreamFactory",
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:stream_data, "~> 0.5"},
      {:ecto_sql, "~> 3.0", optional: true},
      {:postgrex, ">= 0.0.0", only: :test},
      {:ex_doc, "~> 0.28", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: :dev, runtime: false}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp aliases do
    [
      test: ["ecto.create --quiet", "ecto.migrate", "test"]
    ]
  end

  defp docs do
    [
      source_url: @project_url,
      source_ref: "v#{@version}",
      extras: ["README.md", "CHANGELOG.md"],
      main: "readme"
    ]
  end

  defp package do
    [
      maintainers: ["Igor Barchenkov"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => @project_url,
        "Documentation" => "https://hexdocs.pm/ecto_stream_factory",
        "Changelog" => "https://hexdocs.pm/ecto_stream_factory/changelog.html"
      }
    ]
  end
end


lib/ecto_stream_factory/factory.ex
defmodule EctoStreamFactory.Factory do
  @moduledoc false

  def build(module, generator_name, attrs) do
    do_build(module, generator_name, attrs, &build_list/4)
  end

  def build!(module, generator_name, attrs) do
    do_build(module, generator_name, attrs, &build_list!/4)
  end

  defp do_build(module, generator_name, attrs, build_list_fun) do
    module
    |> build_list_fun.(1, generator_name, attrs)
    |> hd()
  end

  def build_list(module, count, generator_name, attrs) do
    do_build_list(module, count, generator_name, attrs, &create_records_stream/3)
  end

  def build_list!(module, count, generator_name, attrs) do
    do_build_list(module, count, generator_name, attrs, &create_records_stream!/3)
  end

  defp do_build_list(module, count, generator_name, attrs, create_records_stream_fun) do
    module
    |> create_records_stream_fun.(generator_name, attrs)
    |> Enum.take(count)
  end

  def insert(module, repo, generator_name, attrs, opts) do
    do_insert(module, repo, generator_name, attrs, opts, &insert_list/6)
  end

  def insert!(module, repo, generator_name, attrs, opts) do
    do_insert(module, repo, generator_name, attrs, opts, &insert_list!/6)
  end

  def do_insert(module, repo, generator_name, attrs, opts, insert_list_fun) do
    module
    |> insert_list_fun.(repo, 1, generator_name, attrs, opts)
    |> hd()
  end

  def insert_list(module, repo, count, generator_name, attrs, opts) do
    do_insert_list(module, repo, count, generator_name, attrs, opts, &create_records_stream/3)
  end

  def insert_list!(module, repo, count, generator_name, attrs, opts) do
    do_insert_list(module, repo, count, generator_name, attrs, opts, &create_records_stream!/3)
  end

  def do_insert_list(module, repo, count, generator_name, attrs, opts, create_records_stream_fun) do
    module
    |> create_records_stream_fun.(generator_name, attrs)
    |> Stream.map(&insert_record(&1, repo, opts))
    |> Enum.take(count)
  end

  defp create_records_stream(module, generator_name, attrs) do
    do_create_records_stream(module, generator_name, attrs, &merge_attributes/4)
  end

  defp create_records_stream!(module, generator_name, attrs) do
    do_create_records_stream(module, generator_name, attrs, &merge_attributes!/4)
  end

  defp do_create_records_stream(module, generator_name, attrs, merge_attrs_fun) do
    module
    |> create_stream(generator_name)
    |> Stream.with_index(1)
    |> Stream.map(&merge_attrs_fun.(&1, Map.new(attrs), module, generator_name))
  end

  defp merge_attributes({record, index}, attrs, _, _) when is_struct(record) do
    struct(record, prepare_attributes(attrs, index))
  end

  defp merge_attributes({map, index}, attrs, _, _) when is_map(map) do
    attrs = attrs |> prepare_attributes(index) |> Map.new()
    # refactor to `then/2` when we require Elixir 1.12
    Map.merge(map, attrs)
  end

  defp merge_attributes({maybe_keyword, index}, attrs, _, _) do
    if Keyword.keyword?(maybe_keyword) do
      Keyword.merge(maybe_keyword, prepare_attributes(attrs, index))
    else
      maybe_keyword
    end
  end

  defp merge_attributes!({record, index}, attrs, module, generator_name) when is_struct(record) do
    try do
      struct!(record, prepare_attributes(attrs, index))
    rescue
      err in KeyError ->
        raise EctoStreamFactory.MissingKeyError,
          module: module,
          generator_name: generator_name,
          key: err.key
    end
  end

  defp merge_attributes!({map, index}, attrs, module, generator_name) when is_map(map) do
    check_missing_key(map, attrs, &Map.has_key?/2, module, generator_name)

    merge_attributes({map, index}, attrs, module, generator_name)
  end

  defp merge_attributes!({maybe_keyword, index}, attrs, module, generator_name) do
    if Keyword.keyword?(maybe_keyword) do
      check_missing_key(maybe_keyword, attrs, &Keyword.has_key?/2, module, generator_name)
    end

    merge_attributes({maybe_keyword, index}, attrs, module, generator_name)
  end

  defp check_missing_key(map_or_keyword, attrs, has_key_fun, module, generator_name) do
    missing_key =
      attrs
      |> Map.keys()
      |> Enum.find_value(fn key ->
        if has_key_fun.(map_or_keyword, key) do
          false
        else
          key
        end
      end)

    if missing_key do
      raise EctoStreamFactory.MissingKeyError,
        module: module,
        generator_name: generator_name,
        key: missing_key
    end
  end

  defp prepare_attributes(attrs, index) do
    Enum.map(attrs, fn
      {key, fun} when is_function(fun, 1) ->
        {key, fun.(index)}

      {_key, _val} = result ->
        result
    end)
  end

  defp insert_record(record, repo, opts) do
    apply(repo, :insert!, [record, opts])
  end

  defp create_stream(module, generator_name) do
    function_name = function_from_generator_name(generator_name)

    if function_exported?(module, function_name, 0) do
      apply(module, function_name, [])
    else
      raise EctoStreamFactory.UndefinedGeneratorError,
        module: module,
        generator_name: generator_name
    end
  end

  def function_from_generator_name(generator_name) when is_atom(generator_name) do
    generator_name |> Atom.to_string() |> function_from_generator_name()
  end

  def function_from_generator_name(generator_name) when is_binary(generator_name) do
    generator_name |> Kernel.<>("_generator") |> String.to_atom()
  end

  def get_repo(module, opts) do
    case Keyword.fetch(opts, :repo) do
      {:ok, repo} -> repo
      _ -> raise EctoStreamFactory.RepoNotSpecifiedError, module: module
    end
  end
end


lib/ecto_stream_factory/exceptions.ex
defmodule EctoStreamFactory.UndefinedGeneratorError do
  defexception [:module, :generator_name]

  @impl true
  def message(%{module: module, generator_name: gen_name}) do
    """
    No generator defined for #{inspect(gen_name)}.
    Please check for typos or define your generator:

        defmodule #{inspect(module)} do
          def #{EctoStreamFactory.Factory.function_from_generator_name(gen_name)} do
            ...
          end
        end
    """
  end
end

defmodule EctoStreamFactory.RepoNotSpecifiedError do
  defexception [:module]

  @impl true
  def message(%{module: module}) do
    repo_module = module |> String.split(".") |> hd()

    """
    :repo option is missing. Please configure the factory:

        defmodule #{inspect(module)} do
          use EctoStreamFactory, repo: #{repo_module}.Repo
          ...
        end
    """
  end
end

defmodule EctoStreamFactory.MissingKeyError do
  defexception [:module, :generator_name, :key]

  @impl true
  def message(%{module: module, generator_name: gen_name, key: key}) do
    """
    #{inspect(module)}.#{EctoStreamFactory.Factory.function_from_generator_name(gen_name)} does not generate #{inspect(key)} field.
    """
  end
end


lib/ecto_stream_factory/ecto_stream_factory.ex
defmodule EctoStreamFactory do
  @moduledoc """
  Defines a factory that can be used in tests and database seeds files.

  ## Examples

      defmodule MyApp.Factory do
        use EctoStreamFactory, repo: MyApp.Repo

        def user_generator do
          gen all name <- string(:alphanumeric, min_length: 1),
                  age <- integer(15..80) do
            %User{name: name, age: age}
          end
        end

        def post_generator do
          gen all author <- user_generator(),
                  body <- string(:alphanumeric, min_length: 10) do
            %Post{author: author, body: body}
          end
        end
      end
  """

  alias EctoStreamFactory.Factory
  alias Ecto.Schema

  defmacro __using__(opts) do
    quote bind_quoted: [opts: opts] do
      use ExUnitProperties

      @repo Factory.get_repo(__MODULE__, opts)

      defguardp is_valid_generator_name(value) when is_atom(value) or is_binary(value)
      defguardp is_valid_amount(value) when is_integer(value) and value > 0

      def build(generator_name, attrs \\ []) when is_valid_generator_name(generator_name) do
        Factory.build(__MODULE__, generator_name, attrs)
      end

      def build!(generator_name, attrs \\ []) when is_valid_generator_name(generator_name) do
        Factory.build!(__MODULE__, generator_name, attrs)
      end

      def build_list(amount, generator_name, attrs \\ [])
          when is_valid_generator_name(generator_name) and
                 is_valid_amount(amount) do
        Factory.build_list(__MODULE__, amount, generator_name, attrs)
      end

      def build_list!(amount, generator_name, attrs \\ [])
          when is_valid_generator_name(generator_name) and
                 is_valid_amount(amount) do
        Factory.build_list!(__MODULE__, amount, generator_name, attrs)
      end

      def insert(generator_name, attrs \\ [], opts \\ [])
          when is_valid_generator_name(generator_name) do
        Factory.insert(__MODULE__, @repo, generator_name, attrs, opts)
      end

      def insert!(generator_name, attrs \\ [], opts \\ [])
          when is_valid_generator_name(generator_name) do
        Factory.insert!(__MODULE__, @repo, generator_name, attrs, opts)
      end

      def insert_list(amount, generator_name, attrs \\ [], opts \\ [])
          when is_valid_generator_name(generator_name) and is_valid_amount(amount) do
        Factory.insert_list(__MODULE__, @repo, amount, generator_name, attrs, opts)
      end

      def insert_list!(amount, generator_name, attrs \\ [], opts \\ [])
          when is_valid_generator_name(generator_name) and is_valid_amount(amount) do
        Factory.insert_list!(__MODULE__, @repo, amount, generator_name, attrs, opts)
      end
    end
  end

  @type generator_name() :: atom() | String.t()

  @type overwrite_value :: term() | (pos_integer() -> term())

  @typedoc """
  Keyword list or map where values can be of any literal types or one arity functions that generate sequential data.
  """
  @type overwrites() :: [{atom(), overwrite_value()}] | %{term() => overwrite_value()}

  @type amount() :: pos_integer()

  @typedoc """
  Same as options for `c:Ecto.Repo.insert/2`.
  """
  @type insert_opts() :: Keyword.t()

  @doc """
  Takes a single instance from a StreamData [gen/1](`ExUnitProperties.gen/1`) generator.

  If the generator returs a struct, map or keyword list, it also merges the result with `overwrites`.

  ## Examples

      iex> Factory.build(:user)
      %User{id: nil, name: "a", age: 33}

      iex> Factory.build(:user, name: "foo")
      %User{id: nil, name: "foo", age: 49}
  """
  @callback build(generator_name(), overwrites()) :: term()

  @doc """
  Same as `c:build/2`, but raises an error if some key in `overwrites` does not exist in the generated struct, map or keyword list.

  ## Examples

      iex> Factory.build!(:user)
      %User{id: nil, name: "b", age: 28}

      iex> Factory.build!(:user, name: "foo", weight: 101)
      ** (EctoStreamFactory.MissingKeyError) MyApp.Factory.user_generator does not generate :weight field.
  """
  @doc since: "0.2.0"
  @callback build!(generator_name(), overwrites()) :: term()

  @doc ~S"""
  Same as `c:build/2`, but instantiates a list of structs.

  ## Examples

      iex> Factory.build_list(2, :user, name: fn n -> "user#{n}" end)
      [%User{id: nil, name: "user1"}, %User{id: nil, name: "user2"}]
  """
  @callback build_list(amount(), generator_name(), overwrites) :: nonempty_list(term())

  @doc ~S"""
  Same as `c:build_list/3`, but raises an error if some key in `overwites` does not exist in the generated entities.
  """
  @doc since: "0.2.0"
  @callback build_list!(amount(), generator_name(), overwrites) :: nonempty_list(term())

  @doc """
  Similar to `c:build/2`, but expects a generator to return an Ecto.Schema struct to insert it into the database.

  ## Examples

      iex> Factory.insert(:user, [email: "duplicated@example.com"], on_conflict: :nothing)
      %User{id: 2, email: "duplicated@example.com"}
  """
  @callback insert(generator_name(), overwrites(), insert_opts()) :: Schema.t()

  @doc """
  Same as `c:insert/3`, but raises an error if some key in `overwites` does not exist in the generated struct.
  """
  @doc since: "0.2.0"
  @callback insert!(generator_name(), overwrites(), insert_opts()) :: Schema.t()

  @doc """
  Same as `c:insert/3`, but inserts a list of structs.

  ## Examples

      iex> Factory.insert_list(3, :user, role: "admin")
      [%User{id: 3, role: "admin"}, %User{id: 4, role: "admin"}, %User{id: 5, role: "admin"}]
  """
  @callback insert_list(amount(), generator_name(), overwrites(), insert_opts()) ::
              nonempty_list(Schema.t())

  @doc ~S"""
  Same as `c:insert_list/4`, but raises an error if some key in `overwites` does not exist in the generated structs.
  """
  @doc since: "0.2.0"
  @callback insert_list!(amount(), generator_name(), overwrites(), insert_opts()) ::
              nonempty_list(Schema.t())
end
